kani-driver/src/project.rs

```diff
@@ -0,0 +1,282 @@
+    /// If this flag is true, there should only be up to one artifact of a
+    merged_artifacts: bool,
+}
+
+impl Project {
+    /// Get all harnesses from a project. This will include all test and proof harnesses.
+    /// We could create a `get_proof_harnesses` and a `get_tests_harnesses` later if we see the
+    /// need to split them.
+    pub fn get_all_harnesses(&self) -> Vec<&HarnessMetadata> {
+        self.metadata
+            .iter()
+            .flat_map(|crate_metadata| {
+                crate_metadata.proof_harnesses.iter().chain(crate_metadata.test_harnesses.iter())
+            })
+            .collect()
+    }
+
+    /// Return the matching artifact for the given harness.
+    ///
+    /// If the harness has information about the goto_file we can use that to find the exact file.
+    /// For cases where there is no goto_file, we just assume that everything has been linked
+    /// together. I.e.: There should only be one artifact of the given type.
+    pub fn get_harness_artifact(
+        &self,
+        harness: &HarnessMetadata,
+        typ: ArtifactType,
+    ) -> Option<&Artifact> {
+        let expected_path = if self.merged_artifacts {
+            None
+        } else {
+            harness.goto_file.as_ref().map(|goto_file| convert_type(goto_file, SymTabGoto, typ))
+        };
+        trace!(?harness.goto_file, ?expected_path, ?typ, "get_harness_artifact");
+        self.artifacts.iter().find(|artifact| {
+            artifact.has_type(typ)
+                && expected_path.as_ref().map_or(true, |goto_file| *goto_file == artifact.path)
+        })
+    }
+}
+
+/// Information about a build artifact.
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct Artifact {
+    /// The path for this artifact in the canonical form.
+    path: PathBuf,
+    /// The type of artifact.
+    typ: ArtifactType,
+}
+
+impl AsRef<Path> for Artifact {
+    fn as_ref(&self) -> &Path {
+        self.path.as_ref()
+    }
+}
+
+impl Deref for Artifact {
+    type Target = Path;
+    fn deref(&self) -> &Self::Target {
+        &self.path
+    }
+}
+
+impl Artifact {
+    /// Create a new existing artifact.
+    /// This will fail if the artifact doesn't exist.
+    fn try_new(path: &Path, typ: ArtifactType) -> Result<Self> {
+        Ok(Artifact { path: path.canonicalize()?, typ })
+    }
+
+    fn has_type(&self, typ: ArtifactType) -> bool {
+        self.typ == typ
+    }
+}
+
+fn cargo_artifact(metadata: &Path, typ: ArtifactType) -> Option<Artifact> {
+    let path = convert_type(metadata, Metadata, typ);
+    if path.exists() { Artifact::try_new(&path, typ).ok() } else { None }
+}
+
+fn dump_metadata(metadata: &KaniMetadata, path: &Path) {
+    let out_file = File::create(path).unwrap();
+    let writer = BufWriter::new(out_file);
+    serde_json::to_writer_pretty(writer, &metadata).unwrap();
+}
+
+pub fn cargo_project(session: &KaniSession) -> Result<Project> {
```

Comment:
My main reason was simplicity and trying really hard to scope down this PR. I wanted to try to avoid changing `call_cargo.rs` because that code itself needs some refactoring (and this PR was already getting too big). So this code does a simple mapping between what `call_cargo.rs` generate.

FYI, the refactoring I'm talking about is to be able to get rid of cleaning the build artifacts.